ADAL/src/ADEnrollmentGateway.m

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADEnrollmentGateway.h"
+
+// Keys for Intune Enrollment ID
+NSString* const ADIntuneAppProtectionEnrollmentID = @"intune_app_protection_enrollment_id_V";
+NSString* const ADIntuneAppProtectionEnrollmentIDVersion = @"1";
+#define AD_INTUNE_ENROLLMENT_ID_KEY [ADIntuneAppProtectionEnrollmentID stringByAppendingString:ADIntuneAppProtectionEnrollmentIDVersion]
+NSString* const enrollmentIdArray = @"enrollment_ids";
+
+NSString* const tid = @"tid";
+NSString* const oid = @"oid";
+NSString* const unique_account_id = @"unique_account_id";
+NSString* const user_id = @"user_id";
+NSString* const enrollment_id = @"enrollment_id";
+
+// Keys for Intune Resource
+NSString* const ADIntuneAppProtectionResourceID = @"intune_mam_resource_V";
+NSString* const ADIntuneAppProtectionResourceIDVersion = @"1";
+#define AD_INTUNE_RESOURCE_ID_KEY [ADIntuneAppProtectionEnrollmentID stringByAppendingString:ADIntuneAppProtectionEnrollmentIDVersion]
+
+static NSString* ADIntuneEnrollmentIdJSON = nil;
+static NSString* ADIntuneResourceJSON = nil;
+
+
+@interface ADEnrollmentGateway()
+
++ (NSString*) getEnrollmentIDForIdentifier:(BOOL (^)(NSDictionary*)) idBlock;
++ (NSString *) normalizeAuthority:(NSString *)authority;
+
+@end
+
+@implementation ADEnrollmentGateway
+
++ (NSString*) getEnrollmentIDForIdentifier:(BOOL (^)(NSDictionary*)) idBlock
+{
+    NSString* enrollIdJSON = [ADEnrollmentGateway allEnrollmentIds];
+
+    if (!enrollIdJSON)
+        return nil;
+
+    NSError* error = nil;
+    id enrollIds = [NSJSONSerialization JSONObjectWithData:[enrollIdJSON dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&error];
+
+    if (error)
+        return nil;
+
+    enrollIds = enrollIds[enrollmentIdArray];
```

Comment:
please also validate that enrollIds is valid and that enrollIds is actually a dictionary, because any app can actually write to user defaults and can corrupt them.